File: Database_IMGVR/02.make_vOTU_representatives_list.pl
#!/usr/bin/perl

use strict;
use warnings;

# Aim: make vOTU representative list; pick the high quality and long phage genome

# Store all genome property
my %Gn_property = (); # $gn => [0] $vOTU, [1] $length, [2] $completeness, [3] $genome_quality, [4] $tax
my %VOTU2gn = (); # $vOTU => $gn separated by "\t";
open IN, "IMGVR_all_Sequence_information.tsv";
while (<IN>){
	chomp;
	if (!/^#/){
		my @tmp = split (/\t/);
		my $gn = $tmp[0];
		my $vOTU = $tmp[5];
		my $length = $tmp[6];
		my $completeness = $tmp[8];
		my $genome_quality = $tmp[9];
		my $tax = $tmp[11];
		$Gn_property{$gn}[0] = $vOTU;
		$Gn_property{$gn}[1] = $length;
		$Gn_property{$gn}[2] = $completeness;
		$Gn_property{$gn}[3] = $genome_quality;
		$Gn_property{$gn}[4] = $tax;
		if (!exists $VOTU2gn{$vOTU}){
			$VOTU2gn{$vOTU} = $gn;
		}else{
			$VOTU2gn{$vOTU} .= "\t".$gn;
		}
	}
}
close IN;

# Pick vOTU representative 
my %VOTU2rep = (); # $vOTU => $gn_rep (representative genome)
foreach my $vOTU (sort keys %VOTU2gn){
	my @Gn = split (/\t/, $VOTU2gn{$vOTU});
	
	# Divide genomes into 4 collections by genome quality
	my @Gn2 = (); # 2nd level genome collection for picking, containing "Reference"
	my @Gn3 = (); # 3rd level genome collection for picking, containing "High-quality"
	my @Gn4 = (); # 4th level genome collection for picking, containing "Genome fragment"
	my @Gn5 = (); # 5th level genome collection for picking, containing "Unsure (completeness > 120%)" or "Unsure (Ns)" or "Concatemer_artifact" 
	foreach my $gn (@Gn){
		if ($Gn_property{$gn}[3] eq "Reference"){
			push @Gn2, $gn;
		}elsif ($Gn_property{$gn}[3] eq "High-quality"){
			push @Gn3, $gn;
		}elsif ($Gn_property{$gn}[3] eq "Genome fragment"){
			push @Gn4, $gn;
		}elsif ($Gn_property{$gn}[3] =~ /Unsure/ or $Gn_property{$gn}[3] eq "Concatemer_artifact"){
			push @Gn5, $gn;
		}
	}
	
	if (@Gn2){ # If the 2nd level genome collection is not empty
		my $gn_rep = $Gn2[0]; 
		for(my $i=1; $i<=$#Gn2; $i++){
			if ($Gn_property{$Gn2[$i]}[1] > $Gn_property{$gn_rep}[1]){ # Pick the longest one from the 2nd level genome collection
				$gn_rep = $Gn2[$i];
			}
		}
		$VOTU2rep{$vOTU} = $gn_rep;
	}elsif (!@Gn2 and @Gn3){ # If the 2nd level genome collection is empty, and 3rd level genome collection is not empty
		my $gn_rep = $Gn3[0]; 
		for(my $i=1; $i<=$#Gn3; $i++){
			if ($Gn_property{$Gn3[$i]}[1] > $Gn_property{$gn_rep}[1]){ # Pick the longest one from the 3nd level genome collection
				$gn_rep = $Gn3[$i];
			}
		}
		$VOTU2rep{$vOTU} = $gn_rep;
	}elsif (!@Gn2 and !@Gn3 and @Gn4){ # If the 3rd level genome collection is empty, and 4th level genome collection is not empty
		my $gn_rep = $Gn4[0]; 
		for(my $i=1; $i<=$#Gn4; $i++){
			if ($Gn_property{$Gn4[$i]}[1] > $Gn_property{$gn_rep}[1]){ # Pick the longest one from the 4th level genome collection
				$gn_rep = $Gn4[$i];
			}
		}
		$VOTU2rep{$vOTU} = $gn_rep;
	}elsif (!@Gn2 and !@Gn3 and !@Gn4 and @Gn5){ # If the 4th level genome collection is empty, and 5th level genome collection is not empty
		my $gn_rep = $Gn5[0]; 
		for(my $i=1; $i<=$#Gn5; $i++){
			if ($Gn_property{$Gn5[$i]}[1] > $Gn_property{$gn_rep}[1]){ # Pick the longest one from the 5th level genome collection
				$gn_rep = $Gn5[$i];
			}
		}
		$VOTU2rep{$vOTU} = $gn_rep;		
	}
}

# Write down vOTU representatives table 

open OUT, ">IMGVR_all_phage_vOTU_representatives.txt";
print OUT "vOTU\tvOTU representative\tvOTU representative taxonomy\n";
foreach my $key (sort keys %VOTU2rep){
	print OUT "$key\t$VOTU2rep{$key}\t$Gn_property{$VOTU2rep{$key}}[4]\n";
}
close OUT;
close OUT;


# Make a new folder containing vOTU representatives
`mkdir vOTU_representatives_phage_genomes`;

open OUT, ">tmp.copy_genomes.sh";
foreach my $vOTU (sort keys %VOTU2rep){
	my $gn_rep = $VOTU2rep{$vOTU};
	
	# Replace "|" to "__" in the gn name
	my $gn_rep_new = $gn_rep; $gn_rep_new =~ s/\|/\_\_/g;
	print OUT "cp Individual_phage_genomes/$gn_rep_new.fasta vOTU_representatives_phage_genomes/$vOTU\_\_\_$gn_rep_new.fasta\n";
}
close OUT;

`cat tmp.copy_genomes.sh | parallel -j 20`;

`rm tmp.copy_genomes.sh`;

File: Database_IMGVR/02.make_vOTU_representatives_list.py
#!/usr/bin/env python3

try:
    import warnings
    import sys
    import os
    warnings.filterwarnings("ignore")
    from pathlib import Path
    from collections import defaultdict    
except Exception as e:
    sys.stderr.write(str(e) + "\n\n")
    exit(1)

# Aim: make vOTU representative list; pick the high quality and long virus genome

# Step 1 Store all genome property
Gn_property = {} # $gn => [vOTU, length, completeness, genome_quality, tax]
vOTU2gns = defaultdict(list) # $vOTU => [gns]
with open("IMGVR_all_Sequence_information.tsv", 'r') as IN:
    for line in IN:
        if not line.startswith('UVIG'):
            tmp = line.rstrip('\n').split('\t')
            gn = tmp[0]
            vOTU = tmp[5]
            length = tmp[6]
            completeness = tmp[10]
            genome_quality = tmp[12]
            tax = tmp[14]
            Gn_property[gn] = [vOTU, length, completeness, genome_quality, tax]
            vOTU2gns[vOTU].append(gn) # Add gn list in each vOTU
IN.close()            

# Step 2 Pick vOTU representative 
vOTU2rep = {} # vOTU => gn_rep (representative genome)
for vOTU in sorted(vOTU2gns.keys()):
    Gns = vOTU2gns[vOTU] # The list that contains all gns in this vOTU
    
    # Divide genomes into 4 collections by genome quality
    Gns2 = [] # 2nd level genome collection for picking, containing "Reference"
    Gns3 = [] # 3rd level genome collection for picking, containing "High-quality"
    Gns4 = [] # 4th level genome collection for picking, containing "Genome fragment"
    Gns5 = [] # 5th level genome collection for picking, containing "Unsure" (>120% or no completeness estimate) 
    for gn in Gns:
        if Gn_property[gn][3] == "Reference":
            Gns2.append(gn)
        elif Gn_property[gn][3] == "High-quality":
            Gns3.append(gn)
        elif Gn_property[gn][3] == "Genome fragment":
            Gns4.append(gn)
        elif Gn_property[gn][3] == "Unsure":
            Gns5.append(gn)

    if Gns2: # If the 2nd level genome collection is not empty
        gn_rep = Gns2[0]
        for i in range(1, len(Gns2)):
            if int(Gn_property[Gns2[i]][1]) > int(Gn_property[gn_rep][1]): # Pick the longest one from the 2nd level genome collection
                gn_rep = Gns2[i]
        vOTU2rep[vOTU] = gn_rep
    elif not Gns2 and Gns3: # If the 2nd level genome collection is empty and the 3rd level genome collection is not empty
        gn_rep = Gns3[0]
        for i in range(1, len(Gns3)):
            if int(Gn_property[Gns3[i]][1]) > int(Gn_property[gn_rep][1]): # Pick the longest one from the 3rd level genome collection
                gn_rep = Gns3[i]
        vOTU2rep[vOTU] = gn_rep
    elif not Gns2 and not Gns3 and Gns4: # If the 2nd and 3rd level genome collection are empty and the 4th level genome collection is not empty
        gn_rep = Gns4[0]
        for i in range(1, len(Gns4)):
            if int(Gn_property[Gns4[i]][1]) > int(Gn_property[gn_rep][1]): # Pick the longest one from the 4th level genome collection
                gn_rep = Gns4[i]
        vOTU2rep[vOTU] = gn_rep
    elif not Gns2 and not Gns3 and not Gns4 and Gns5: # If all the above genome collections are empty, pick the first one from the 5th level genome collection
        vOTU2rep[vOTU] = Gns5[0]
    else:
        warnings.warn("No genomes found for vOTU: " + vOTU)

# Step 3 Write the vOTU representative list
with open("IMGVR_all_phage_vOTU_representatives.txt", 'w') as OUT:
    OUT.write("vOTU\tvOTU representative\tvOTU representative taxonomy\n")
    for vOTU, gn_rep in vOTU2rep.items():
        OUT.write(vOTU + "\t" + gn_rep + "\t" + Gn_property[gn_rep][4] + "\n")
OUT.close()  

